AMReX: 23.06+ Multi-Dim Support

ax3l opened the pull request:
This updated the Spack package to allow to install AMReX, modules of AMReX in E4S deployments, and dependent packages, with support for multiple dimensions. Due to an [upstream change in AMReX](https://github.com/AMReX-Codes/amrex/pull/3309), we do not longer need to ship three, binary incompatible modules (resulting in three, incompatible HPC modules, resulting in three, incompatible variants per downstream app, etc. ...).

More background in this blog article: https://bssw.io/blog_posts/rethinking-software-variants

ax3l committed:
AMReX: 23.06+ Multi-Dim Support

This updated the Spack package to allow to install AMReX, modules of
AMReX in E4S deployments and dependent packages with support for
multiple dimensions. Due to an upstream change in AMReX, we do not
longer need to ship three, binary incompatible package variants.

ax3l commented:
I think I found a bug in `variant(when=...` - it always takes the default of the newer one, even if it is not applicable.
Mini-demo: 

`spack install amrex dimension=1,2,3` should:
- either pick `develop` or
- fail with Spack error
```
==> Error: multiple values are not allowed for variant "dimensions" in package "amrex"
```

But instead it tries to use `23.05`, which is clearly not yet a multi variant, and fails during CMake configure (as it should).

ax3l commented:
Looks like `amrex@develop` does the right thing:
```
$ spack spec amrex@develop
Concretized
--------------------------------
amrex@develop ... dimensions=1,2,3 ...
```

But not for older versions:
```
$ spack spec amrex

Concretized
--------------------------------
amrex@23.05 ... dimensions=1,2,3
```
```
$ spack spec amrex@23.04

Concretized
--------------------------------
amrex@23.04 ... dimensions=1,2,3
```

ax3l commented:
Pushing the latest AMReX now, so this Spack bug will only will be weird for <=`23.05`.

ax3l commented:
@spack/e4s do you understand why the `e4s-oneapi-build` CI builder builds AMReX `22.12` instead of the latest version?

eugeneswalker commented:
> @spack/e4s do you understand why the `e4s-oneapi-build` CI builder builds AMReX `23.06` (latest) _and_ `22.12` instead of only the latest version `23.06`?
> 
> I cannot find where this CI runner is configured to build `22.12`, but in order to fix it please add `dimensions=3` for <=`23.05` releases of AMReX to enforce the old default, which this spack bug #38302 prevents to auto-pick.

Can you re-base this on develop? We updated the OneAPI CI stack on develop -- both the runner image and the Spack environment -- and if you re-base on develop, the older `amrex` will come somehow from the concretizer making a decision about what version to use for `amrex +sycl`. We can (1) re-base on develop and (2) simply add `amrex +sycl dimensions=3` to the relevant spec line and it should be ok. What do you think?

ax3l commented:
Rebased two weeks ago, shall I rebase again? :)

eugeneswalker commented:
It seems like the restriction of multi-dim to versions `@23.06:` is not working. The E4S OneAPI CI env is concretizing `amrex +sycl` to `dimensions=1,2,3`:
```
==> Concretized amrex+sycl
 -   o4vmisi  amrex@22.12%oneapi@2023.1.0~amrdata~cuda~eb~fortran~hdf5~hypre~ipo+linear_solvers+mpi~openmp~particles~petsc~pic~plotfile_tools~rocm~shared~sundials+sycl~tiny_profile build_system=cmake build_type=Release dimensions=1,2,3 generator=make precision=double arch=linux-ubuntu20.04-x86_64
```

Causing the obv error:
```
1 error found in build log:
     12    -- Check for working CXX compiler: /spack/lib/spack/env/oneapi/icpx - skipped
     13    -- Detecting CXX compile features
     14    -- Detecting CXX compile features - done
     15    -- CMake version: 3.26.3
     16    -- AMReX installation directory: /spack/opt/spack/linux-ubuntu20.04-x86_64/oneapi-2023.1.0/amrex-22.12-o4vmisib3mhy7tqrzmlthmnk7ektvzao
     17    -- Build type set by user to 'Release'.
  >> 18    CMake Error at Tools/CMake/AMReXOptions.cmake:28 (message):
     19      AMReX_SPACEDIM=1;2;3 is not allowed.  Must be one of 1;2;3
     20    Call Stack (most recent call first):
     21      CMakeLists.txt:84 (include)
     22
     23
     24    -- Configuring incomplete, errors occurred!
```

We can workaround this by appending a commit to this PR that explicitly specifies `amrex +sycl dimensions=3` explicit in the E4S OneAPI CI Env:

Change this line:
https://github.com/spack/spack/blob/develop/share/spack/gitlab/cloud_pipelines/stacks/e4s-oneapi/spack.yaml#L229

To say `amrex +sycl dimensions=3` and we will be good

ax3l committed:
[E4S] oneAPI AMReX < 23.06 Variant

Work-around the auto-concretization to the multi-dim of `dimensions`,
which only in 23.06+ became a multi-variant.

ax3l commented:
Thank you @eugeneswalker, rebased and pushed accordingly.

eugeneswalker commented:
The only failing CI job now is in `E4S Cray RHEL` CI stack w/ `%cce@15` 
* [amrex/2uhnojf 23.07 cce@=15.0.1 linux-rhel8-zen4 E4S Cray](https://gitlab.spack.io/spack/spack/-/jobs/7745604)

Unfortunately, the error trace is not very informative. I will try the build on this machine manually and post update here when I see what the problem is.

eugeneswalker commented:
@spackbot run pipeline

eugeneswalker commented:
> The only failing CI job now is in `E4S Cray RHEL` CI stack w/ `%cce@15`
> 
>     * [amrex/2uhnojf 23.07 cce@=15.0.1 linux-rhel8-zen4 E4S Cray](https://gitlab.spack.io/spack/spack/-/jobs/7745604)
> 
> 
> Unfortunately, the error trace is not very informative. I will try the build on this machine manually and post update here when I see what the problem is.

Ran this locally and the build completed just fine. Not sure what the issue is but I've restarted the pipeline. Let us see if we get a different result the second try...

eugeneswalker committed:
e4s cray rhel ci: temporarily disable amrex build until spurious ci failure can be resolved

eugeneswalker commented:
Happy to merge whenever you think it is best @ax3l
